### ISSUE TITLE
Achieving 200ms latency

### ISSUE THREAD

**Issue by alpha-adam** [canopyai/Orpheus-TTS-0.1#222]
How can I achieve the advertised 200ms latency? I've tried to export to GPTQ, but still getting the same latency of around 900ms on a H100. Cheers.

**Comment by amuvarma13** [comment]
Is snac on gpu?

**Comment by TB360** [comment]
> How can I achieve the advertised 200ms latency? I've tried to export to GPTQ, but still getting the same latency of around 900ms on a H100. Cheers.

question is if you wanna get 200 ms on TTFT or to all the process completed :)
You do streaming or one batch? this is also important to describe. 

Dont know your setup but with torch.compile you can achieve almost all the speed you need. Regards

**Comment by amuvarma13** [comment]
200 ttfbyte of audio. This using vllm streaming async engine, but you can probably get better on tensorrtllm.

**Comment by alpha-adam** [comment]
Sorry for the late reply @amuvarma13, I made a silly mistake. I had forgotten that vllm requires a warm up run before the TTFT is optimised. When I re tested, I got ~250ms in batch mode on an H100. @TB360 Thank you for your reply as well. I do mean TTFT, I am interested in real time voice applications.

**Comment by rvt123** [comment]
@alpha-adam, can you describe your steps for the best possible latency?

**Comment by alpha-adam** [comment]
@rvt123 Sure.

- Use at least a H100 if you want ~200ms
- Run a warm up test first as vLLM is slow on the first request after model load
- Ensure snac is on the GPU, it is the default
- Use the Orpheus inference code
- Use the 'ft' model, not the 'pretrained' model
- You can get even better latency if you use trt-llm but it's 10x more work

**Comment by alpha-adam** [comment]
@rvt123 I got it down to 130ms latency x4 concurrent requests on a 48GB A40 with triton inference server using tensor-rt llm. But is was lot's of work

**Comment by rvt123** [comment]
Did you try the SGLang server, this is supposed to be as much as faster as TRT-LLM  but easier to deploy?

**Comment by amuvarma13** [comment]
TRT-LLM is a little faster from our tests - SGlang is still very fast though. You can also get both faster if you do prefix caching with tokens streaming from an LLM periodically as they stream.

**Comment by rvt123** [comment]
Great, did you find SGLang easier to setup ? I am looking for something easier to setup now. Also if you remember and can share, can you share what optimization worked for you in the SGLang? And did you use any quantisations in the models in both TRT-LLM or SGLang, that could help in reducing the latency further without much decrease in the quality?

**Comment by alpha-adam** [comment]
@rvt123 In my testing quantisations don't help with speed for this model, only memory. I have tried sglang before, but I wasn't a fan, felt like more work than vllm and worse performance. Triton is a big step in performance for this model for me.

**Comment by taresh18** [comment]
@alpha-adam Hi, could you please share how to get it running using troton inference server?

**Comment by alpha-adam** [comment]
@taresh18 Sure.

At the core of Orpheus it is essentially just a normal LLM. Meaning you can use a range of tools to perform inference. So step one, is get Triton Inference Server running using tensorrt-llm for the engine. Start by getting it work on the standard llama 3b model. Once you have that working, try performing inference on the orpheus model, note that it will output meaningless tokens because they aren't converted to audio yet. Once you have that working, you need to port the orpheus specific snac decoding portion, refer to their inference code for this orpheus_tts_pypi/orpheus_tts/decoder.py

Then you should be able to use the triton inference server features like batching, and have the benefits of trt-llm faster performance. If you don't need batching you can just use trt-llm directly. But for production triton is highly recommended.

**Comment by amarbayar** [comment]
Hi @alpha-adam - Do you mind sharing which version of tensorrt-llm and Triton you used?

**Comment by alpha-adam** [comment]
@amarbayar 

25.04-trtllm-python-py3TRT-LLM: Bundled with Triton 25.04

nvcr.io/nvidia/tritonserver:25.04-trtllm-python-py3

**Comment by taresh18** [comment]
@alpha-adam  Thanks man. I got it working on tensorrt-llm and getting pretty good latency. Now I am working on handling multiple requests. How did you implement this? The snac decoding part - we need to do this ourselves right? Are you doing this for every request separately? Also did you try trt-llm serve? It seems to be easier to setup than triton inference server. Could you pls share some insights on you achieved this for serving multiple requests? Thanks

**Comment by kadirnar** [comment]
> [@alpha-adam](https://github.com/alpha-adam) Thanks man. I got it working on tensorrt-llm and getting pretty good latency. Now I am working on handling multiple requests. How did you implement this? The snac decoding part - we need to do this ourselves right? Are you doing this for every request separately? Also did you try trt-llm serve? It seems to be easier to setup than triton inference server. Could you pls share some insights on you achieved this for serving multiple requests? Thanks

I tested the H100 system with the vllm library. However, the word "hello" takes 1 second. Were you able to run it in 200 ms?

**Comment by taresh18** [comment]
the first inference will take longer. run some warmup runs for both vllm engine and snac before doing actual inference. you can refer to https://github.com/taresh18/orpheus-streaming

**Comment by alpha-adam** [comment]
@taresh18 You are correct, triton adds a lot of complexity. You can get concurrent streaming requests working with the trt-llm api. For SNAC decoding, it can be a major bottleneck. I didn't look into it deeply, but I suggest running parallel decoders for each request. Comparing pure orpheus and orpheus + snac yields very different results. I adapted the orpheus SNAC decoding for my use case, it doesn't take long, just refer to orpheus_tts_pypi/orpheus_tts/decoder.py

**Comment by shekharmeena2896** [comment]
Hi can you help me, how should I optimise the model that I have fine tuned the hindi one on custom dataset, currently I am getting the latency of 20 sec and more, I want it below 200 ms. I am pretty confused as where to begin

**Comment by amuvarma13** [comment]
> Hi can you help me, how should I optimise the model that I have fine tuned the hindi one on custom dataset, currently I am getting the latency of 20 sec and more, I want it below 200 ms. I am pretty confused as where to begin

The first and biggest issue is that you probably aren't streaming the response out. After that there are different libraries you can run to optimise LLM inference.

**Comment by shekharmeena2896** [comment]
I want to productionize the model, although u have given the streaming code but with my limited gpu's I want an optimal latency. I have read the batten article, they have mentioned tensor rt llm optimisation, that where I am confused, how should I implement it

**Comment by Rikesh0007** [comment]
> I want to productionize the model, although u have given the streaming code but with my limited gpu's I want an optimal latency. I have read the batten article, they have mentioned tensor rt llm optimisation, that where I am confused, how should I implement it

Hello @shekharmeena2896 , Hope this message finds you well. I see you are getting the same challenge that i am facing. I have RTX 4060 TI with 16 GB of vRAM and i am also getting the challenge to keep the inferencing time to nearly ~200ms but it shoots upto 25 seconds. I trained this model "unsloth/orpheus-3b-0.1-ft" using unsloth notebook. Since I am kind of new to this I need help with its setup. If you somehow have sorted this out Would be really helpful if you could help me out. Thank you in advance.

**Comment by odgran** [comment]
Hi, I’m seeing good latency (around 100–200 ms) for sequential inference calls (thanks for the discussion above, helped a lot!), but when I run parallel calls the latency increases significantly. I’ve already tried separating components so that SNAC runs on one GPU and vLLM on another GPU (both H100s). Even with this setup, I can only handle about 8 concurrent calls before latency degrades, despite the decoder GPU still showing ~50% free VRAM (vLLM GPU utilised fully though). Could this be due to GPU contention? If so, what would be the best way to achieve higher concurrency? I couldn’t find documentation on concurrent request handling in the repo (streaming details were helpful, though). Any suggestions or best practices for enabling true concurrent requests would be much appreciated. 

cc: @amuvarma13 @alpha-adam @taresh18

**Comment by alpha-adam** [comment]
@odgran use trt-llm for the model inference, it can handle concurrent requests much better than vllm

**Comment by Rikesh0007** [comment]
Hi @odgran , I have RTX 4060 TI with 16 GB vRAM. It would be really helpful if you could share yours implementation.

**Comment by shekharmeena2896** [comment]
hi @alpha-adam can you please guide me how u did optimisation using tensorrt llm, as the documentation of tensorrt is huge and confusing, I tried but failed.

**Comment by sagaradjay** [comment]
@alpha-adam - thanks for the steps, one query - I am getting a max of 200 tps on a H100 - are the numbers similar to what you see? - I believe that 1000 tps should be very much possible for such a small model like Llama 3B with Triton + TensorRT

**Comment by alpha-adam** [comment]
@sagaradjay I can't remember exactly but 200 sounds correct. However, I would suggest that you look at some of the other options available today. Orpheus is no longer the best LLM based TTS model.

**Comment by raghavtulu-ctrl** [comment]
@alpha-adam, if you could share some insight, I would really love to know which other open-source models do you think are better than Orpheus at the moment, in terms of real time facor, natural sounding voices, memory usage etc ?

**Comment by alpha-adam** [comment]
@raghavtulu-ctrl 

Check this list https://huggingface.co/models?pipeline_tag=text-to-speech&sort=trending

There are many others now that are as small as 0.5B. 

Ironically, I found it much easier and better results to just use Kokoro. The generation time is so fast that streaming is not needed. And it simplifies the whole pipeline. You just need to have a good sentence segmented and queue the outputs. You can get arbitrary length generation with very fast TTFT.

**Comment by sagaradjay** [comment]
@alpha-adam - thanks for the suggestion and confirmation. 200 TPS is quite low IMHO, I was aiming atleast 1000 TPS, the RTF numbers are quite bad and as per my tests - one H100 GPU can have a max concurrency of 8-12 parallel requests (which IMO is too bad).

We tried Kokoro but it looks like the team did not want to release finetuning codes, I need to finetune models for my regional languages and accents.

**Comment by alpha-adam** [comment]
@sagaradjay When you are at the bleeding edge of these technologies it feels like a constant balancing act between important factors like quality, speed, licensing etc. I like to remember that the models are getting better, smaller and more open every month. Building the pipeline and refining it now, and waiting for the models to get better is a useful strategy unless you are already in production.

Unfortunately I don't have any suggestions for other models in particular, I don't have hands on experience with them, I just monitor HF for models and keep up to date on their sizes, quality and architectures.
